bench/sub-async.c

```diff
@@ -0,0 +1,245 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bench.h"
+
+#define NSUBS 500
+#define NMSGS (1 * 1000)
+#define REPEAT 5
+
+typedef struct
+{
+    bool useGlobalDelivery;
+    int maxThreads;
+} benchConfig;
+
+benchConfig configs[] = {
+    {false, 1}, // 1 is not used in this case, just to quiet nats_SetMessageDeliveryPoolSize
+    {true, 1},
+    {true, 2},
+    {true, 3},
+    {true, 4},
+    {true, 5},
+    {true, 6},
+    {true, 7},
+    {true, 8},
+    {true, 9},
+    {true, 10},
+    {true, 11},
+    {true, 12},
+    {true, 13},
+    {true, 14},
+    {true, 15},
+    {true, 16},
+    {true, 17},
+    {true, 18},
+    {true, 19},
+    {true, 20},
+    {true, 23},
+    {true, 31},
+    {true, 47},
+    {true, 100},
+    {true, NSUBS / 2 - 1},
+    {true, NSUBS - 1},
+    {true, NSUBS},
+};
+
+typedef struct
+{
+    natsSubscription *sub;
+    uint64_t sum;
+    uint64_t xor ;
+    uint64_t count;
+    int64_t closedTimestamp;
+} subState;
+
+subState state[NSUBS];
+
+static uint64_t _expectedSum(void)
+{
+    uint64_t sum = 0;
+    for (int64_t i = 0; i < NMSGS; i++)
+        sum += i;
+    return sum;
+}
+
+static uint64_t _expectedXOR(void)
+{
+    uint64_t xor = 0;
+    for (int64_t i = 0; i < NMSGS; i++)
+        xor ^= i;
+    return xor;
+}
+
+static void
+_onMessage(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *closure)
+{
+    subState *s = (subState *)closure;
+    char buf[32];
+    int len = natsMsg_GetDataLength(msg);
+    if (len > 31)
+        len = 31;
+
+    strncpy(buf, natsMsg_GetData(msg), len);
+    buf[len] = '\0';
+    int64_t val = atoi(buf);
+
+    s->sum += val;
+    s->xor ^= val;
+    s->count++;
+
+    natsMsg_Destroy(msg);
+}
+
+static void
+_onComplete(void *closure)
+{
+    subState *s = (subState *)closure;
+    s->closedTimestamp = nats_Now();
+}
+
+static uint64_t expectedSum;
+static uint64_t expectedXOR;
+
+static natsStatus
+_bench(benchConfig *c, int64_t *start, int64_t *end)
+{
+    natsStatus s = nats_Open(-1);
+    natsConnection *conn = NULL;
+    natsOptions *opts = NULL;
+    char buf[16];
+
+    memset(state, 0, sizeof(state));
+    *start = nats_Now();
+
+    IFOK(s, natsOptions_Create(&opts));
+    IFOK(s, nats_SetMessageDeliveryPoolSize(c->maxThreads));
+    IFOK(s, natsOptions_SetErrorHandler(opts, asyncCb, NULL));
+    IFOK(s, natsOptions_UseGlobalMessageDelivery(opts, c->useGlobalDelivery));
+
+    IFOK(s, natsConnection_Connect(&conn, opts));
+
+    for (int i = 0; i < NSUBS; i++)
+    {
+        IFOK(s, natsConnection_Subscribe(&state[i].sub, conn, "foo", _onMessage, &state[i]));
+        IFOK(s, natsSubscription_SetPendingLimits(state[i].sub, INT_MAX, INT_MAX));
```

Comment:
I have seen (in other PRs?) builds on Travis complain about `INT_MAX` being unknown. Regardless, unless you changed something that I missed, setting `-1` for both is the way to set them as unlimited.

Reply:
Please use `-1` instead of `INT_MAX`.